feat: Add InvariantImmiscibleFluid for compositional mixtures

OmarDuran opened the pull request:
This PR adds the InvariantImmiscibleFluid constitutive fluid model for a minimal compositional setup. The fluid is invariant, in the sense of its properties are constant and do not depend on thermodynamic states. While the existing multiphase model can handle similar two-phase immiscible behavior, this model provides a simple setup for the compositional solvers `CompositionalMultiphaseFVM` and `CompositionalMultiphaseHybridFVM`. It is useful for the following run-time checks:


- Minimal configuration for integration tests - no complex thermodynamic parameters
- Allows for fast dynamic tests in large-scale meshes using linear tracer simulations
- Allows for testing and consistency check of FV and MFD discretization on any mesh
- Allows for testing upwinding discretizations under simple gravitational segregation

Minimal setup:
```
<InvariantImmiscibleFluid name="fluid"
                          componentNames="{ H2O, C5H12, CH4 }"
                          phaseNames="{ water, oil, gas }"
                          densities="{ 1000.0, 800.0, 100.0 }"
                          viscosities="{ 0.001, 0.005, 0.00002 }"/>
```

Testing: Phase property validation for both mass/molar basis calculations

OmarDuran committed:
wip: Add invariant immiscible fluid

wip: adding it to cmake list

Update InvariantImmiscibleFluid.cpp

Update InvariantImmiscibleFluid.cpp

wip: updated invariant fluid implementation

Create testInvariantImmiscibleFluid.cpp

Update testInvariantImmiscibleFluid.cpp

Update InvariantImmiscibleFluid.hpp

Update InvariantImmiscibleFluid.cpp

wip: add access methods

Update testInvariantImmiscibleFluid.cpp

Update testInvariantImmiscibleFluid.cpp

Update testInvariantImmiscibleFluid.cpp

wip: testing

Update testInvariantImmiscibleFluid.cpp

wip: adding test

wip: fixing pure abstract methods

Update InvariantImmiscibleFluid.cpp

Update InvariantImmiscibleFluid.cpp

wip: fixing wrapper constructor

wip: test is compiling

wip: running test

wip: running test

wip: uncrustify

wip: register component names

Create invariantImmiscibleFluid_example.xml

wip: rework test

wip: adding driver

OmarDuran committed:
wip: rename input file

OmarDuran committed:
wip: running example

OmarDuran committed:
wip: Hayek test

OmarDuran committed:
Update invariantImmiscibleFluid.xml

OmarDuran committed:
wip: adding phase fractions - composition restriction

OmarDuran committed:
WIP: validation upwinding schemes

OmarDuran committed:
Update invariantImmiscibleFluid.xml

OmarDuran committed:
Merge branch 'develop' into feature/oduran/constant_property_fluid

OmarDuran committed:
wip: Adding MFD and FV setup

OmarDuran committed:
wip: Validated example against semi-analitical solution

OmarDuran committed:
wip: improving unit test

OmarDuran committed:
Update testInvariantImmiscibleFluid.cpp

OmarDuran committed:
Update testInvariantImmiscibleFluid.cpp

OmarDuran committed:
Update invariantImmiscibleFluid_fv.xml

OmarDuran committed:
wip: delete xml files

OmarDuran committed:
wip: delete xml files

OmarDuran committed:
Update testInvariantImmiscibleFluid.cpp

OmarDuran committed:
Update testInvariantImmiscibleFluid.cpp

OmarDuran committed:
Update testInvariantImmiscibleFluid.cpp

OmarDuran committed:
Update InvariantImmiscibleFluid.cpp

OmarDuran committed:
Update InvariantImmiscibleFluid.cpp

OmarDuran committed:
Update InvariantImmiscibleFluid.cpp

OmarDuran committed:
Update InvariantImmiscibleFluid.cpp

paveltomin commented on src/coreComponents/constitutive/fluid/multifluid/constant/InvariantImmiscibleFluid.cpp:
```diff
@@ -0,0 +1,240 @@
+                                       arraySlice1d< real64 const, compflow::USD_COMP - 1 > const & composition ) const
+{
+#ifdef GEOS_USE_DEVICE_CODE
+  // For device code, create slices directly and call compute
```
this does not look standard for geos code and should be probably avoided

paveltomin commented on src/coreComponents/constitutive/fluid/multifluid/constant/InvariantImmiscibleFluid.hpp:
```diff
@@ -0,0 +1,165 @@
+  class KernelWrapper final : public MultiFluidBase::KernelWrapper
+  {
+public:
+    GEOS_HOST_DEVICE
```
better place after the comments

OmarDuran replied:
The decorator was replaced as suggested.

OmarDuran committed:
Merge branch 'develop' into feature/oduran/constant_property_fluid

dkachuma commented on src/coreComponents/constitutive/fluid/multifluid/constant/InvariantImmiscibleFluid.hpp:
```diff
@@ -0,0 +1,165 @@
+  KernelWrapper createKernelWrapper() const;
+
+  GEOS_HOST_DEVICE
+  virtual void update( localIndex const k,
```
The current pattern is to have this method only on the `KernelWrapper`. Then we can call `InvariantImmiscibleFluid::createKernelWrapper::update()`.

OmarDuran replied:
Thank for this clear directive. I made the code compliant with your suggestion.

OmarDuran committed:
wip: removing non-standard code

OmarDuran committed:
Merge branch 'feature/oduran/constant_property_fluid' of https://github.com/GEOS-DEV/GEOS into feature/oduran/constant_property_fluid

dkachuma commented on src/coreComponents/constitutive/fluid/multifluid/constant/InvariantImmiscibleFluid.hpp:
```diff
@@ -0,0 +1,259 @@
+
+  using Deriv = constitutive::multifluid::DerivativeOffset;
+
+  integer nPhase = phaseDensity.value.size();
```
```suggestion
  integer const nPhase = numPhases();
```
Better to rely on this object's properties than caller arguments.

dkachuma commented on src/coreComponents/constitutive/fluid/multifluid/constant/InvariantImmiscibleFluid.hpp:
```diff
@@ -0,0 +1,259 @@
+  using Deriv = constitutive::multifluid::DerivativeOffset;
+
+  integer nPhase = phaseDensity.value.size();
+  integer nComp = phaseCompFraction.value.size( 1 );
```
```suggestion
  integer const nComp = numComponents();
```

dkachuma commented on src/coreComponents/constitutive/fluid/multifluid/constant/InvariantImmiscibleFluid.cpp:
```diff
@@ -0,0 +1,91 @@
+  // check base inputs
+  MultiFluidBase::postInputInitialization();
+
+  integer numPhase = numFluidPhases();
```
I think we have an assumption that number of components equals number of phases. So we need to check `numFluidPhases()` against `numFluidComponents()`.

OmarDuran replied:
The assumption is now checked at the specified location, and the test has been updated accordingly.

dkachuma commented on src/coreComponents/constitutive/fluid/multifluid/constant/InvariantImmiscibleFluid.hpp:
```diff
@@ -0,0 +1,259 @@
+                                                      arraySlice1d< real64 const, compflow::USD_COMP - 1 > const & composition ) const
+{
+  // Create slice views of the member variables at cell k, quadrature point q
+  PhaseProp::SliceType phaseFractionSlice{m_phaseFraction.value[k][q], m_phaseFraction.derivs[k][q]};
```
` m_phaseFraction( k, q )` should give the slice directly. No need to index the values and derivatives separately.

OmarDuran replied:
The slicing is apply directly as you have suggested.

dkachuma reviewed (approved):
Looks good to me.

I however have a feeling that such an incompressible fluid would be quite difficult for the fluid solvers to deal with.

OmarDuran committed:
Update src/coreComponents/constitutive/fluid/multifluid/constant/InvariantImmiscibleFluid.hpp

Co-authored-by: Dickson Kachuma <81433670+dkachuma@users.noreply.github.com>

OmarDuran committed:
Update src/coreComponents/constitutive/fluid/multifluid/constant/InvariantImmiscibleFluid.hpp

Co-authored-by: Dickson Kachuma <81433670+dkachuma@users.noreply.github.com>

OmarDuran committed:
Update InvariantImmiscibleFluid.hpp

OmarDuran committed:
Merge branch 'feature/oduran/constant_property_fluid' of https://github.com/GEOS-DEV/GEOS into feature/oduran/constant_property_fluid

OmarDuran committed:
Update InvariantImmiscibleFluid.cpp

OmarDuran committed:
Update testInvariantImmiscibleFluid.cpp

OmarDuran committed:
Update testInvariantImmiscibleFluid.cpp

OmarDuran committed:
Update testInvariantImmiscibleFluid.cpp

OmarDuran committed:
Merge branch 'develop' into feature/oduran/constant_property_fluid

paveltomin commented:
> I however have a feeling that such an incompressible fluid would be quite difficult for the fluid solvers to deal with.

maybe adding an option to provide phase compressibilities would be nice?

OmarDuran committed:
wip: fix access to make the code CI compliant

Update testInvariantImmiscibleFluid.cpp

wip: fix access to make the code CI compliant

Update InvariantImmiscibleFluid.hpp

Update testInvariantImmiscibleFluid.cpp

Update testInvariantImmiscibleFluid.cpp

Update testInvariantImmiscibleFluid.cpp

OmarDuran committed:
wip: correct conversion for CUDA builds

OmarDuran committed:
wip: explicitly conversion to slice objects

OmarDuran commented:
> Looks good to me.
> 
> I however have a feeling that such an incompressible fluid would be quite difficult for the fluid solvers to deal with.

I can see where incompressible fluids may pose challenges in certain solver setups, especially with mechanics on top. However keep in mind that this zero-order model is intentionally designed to simplify testing and validation within compositional solvers with a minimal setup.

OmarDuran commented:
> > I however have a feeling that such an incompressible fluid would be quite difficult for the fluid solvers to deal with.
> 
> maybe adding an option to provide phase compressibilities would be nice?

Adding optional phase compressibilities is a valid option. However, if needed users can always rely on DeadOilFluid which already include compressibilties via tables.

OmarDuran committed:
wip: explicitly conversion to slice objects

OmarDuran committed:
wip: array-based replaced by StackArray

OmarDuran committed:
wip: StackArray-based approach

OmarDuran committed:
wip: strict type for composition slice

OmarDuran committed:
wip: strict type for composition slice

OmarDuran committed:
Update testInvariantImmiscibleFluid.cpp

OmarDuran committed:
Update testInvariantImmiscibleFluid.cpp

OmarDuran committed:
Update testInvariantImmiscibleFluid.cpp

OmarDuran committed:
Merge branch 'develop' into feature/oduran/constant_property_fluid

OmarDuran committed:
Update testInvariantImmiscibleFluid.cpp

dkachuma commented on src/coreComponents/constitutive/unitTests/testInvariantImmiscibleFluid.cpp:
```diff
@@ -0,0 +1,475 @@
+      composition[2] = 0.3;
+
+      arraySlice1d< real64 const, compflow::USD_COMP - 1 > compositionS( composition );
+      StackArray< real64, 3, 3, multifluid::LAYOUT_PHASE > compositionData( 1, 1, 3 );
+      auto composition_c = compositionData[0][0];
```
```suggestion
      StackArray< real64, 2, 3, compflow::LAYOUT_COMP > compositionData( 1, 3 );
      auto composition_c = compositionData[0];
```
`#include "common/DataLayouts.hpp"`

OmarDuran replied:
Thank you for the suggestion, @dkachuma. Using slices as types in the fluid signatures is indeed non-intuitive. It may be worth a refactor in the near future to simplify the signatures, perhaps using views only.

dkachuma replied:
Agree. The proliferation of these layouts and USDs is not great. We need to simplify/unify this somewhat.

OmarDuran committed:
Update src/coreComponents/constitutive/unitTests/testInvariantImmiscibleFluid.cpp

Co-authored-by: Dickson Kachuma <81433670+dkachuma@users.noreply.github.com>

OmarDuran committed:
wip: make the code compatible with CUDA builds

Update testInvariantImmiscibleFluid.cpp

tjb-ltk reviewed (commented):
An integrated test should be add to validate other components ability to run with this fluid model

OmarDuran committed:
Merge branch 'develop' into feature/oduran/constant_property_fluid

OmarDuran commented:
> An integrated test should be add to validate other components ability to run with this fluid model

This PR is oriented on introducing the invariant fluid model. A set of integration tests is planned and will be added once this  PR is merged.

paveltomin commented:
> > An integrated test should be add to validate other components ability to run with this fluid model
> 
> This PR is oriented on introducing the invariant fluid model. A set of integration tests is planned and will be added once this PR is merged.

don't forget to move it to the merge queue if it ready

tjb-ltk commented:
> > An integrated test should be add to validate other components ability to run with this fluid model
> 
> This PR is oriented on introducing the invariant fluid model. A set of integration tests is planned and will be added once this PR is merged.

Fair enough , didn't realize development branch can be used this way.   What was the rationale for introducing this fluid model instead of modifying the fluid property input of other models to effectively make the prop/ders constant ?

OmarDuran commented:
> > > An integrated test should be add to validate other components ability to run with this fluid model
> > 
> > 
> > This PR is oriented on introducing the invariant fluid model. A set of integration tests is planned and will be added once this PR is merged.
> 
> Fair enough , didn't realize development branch can be used this way. What was the rationale for introducing this fluid model instead of modifying the fluid property input of other models to effectively make the prop/ders constant ?

The rationale is to provide a fluid model that captures zero-order compositional behavior in a minimal setup. This is more robust and transparent than adapting existing models via input hacks, and better suited for testing and validating the compositional flow functionality without having the thermodynamic complexity of the existing models.
